src/tags/object/RichText/domManager.ts

```diff
@@ -0,0 +1,624 @@
+  }
+  createSpans(start: number, end: number):Array<HTMLSpanElement> {
+    const spans: HTMLSpanElement[] = [];
+    const children = [];
+    let wrappableNodes = [];
+
+    for (const node of this.children) {
+      const isTextNode = node instanceof DDTextElement;
+
+      if (node.start >= start && node.end <= end) {
+        wrappableNodes.push(node);
+        continue;
+      }
+      if (wrappableNodes.length) {
+        const spanElement = this.wrapElementsWithSpan(wrappableNodes);
+
+        children.push(spanElement);
+        spans.push(spanElement.node);
+        wrappableNodes = [];
+      }
+      if (start >= node.start && start < node.end || end > node.start && end <= node.end) {
+        if (isTextNode) {
+          const elements = node.createSpanElements(start, end);
+
+          children.push(...elements);
+          spans.push(...elements.filter(el => el instanceof DDSpanElement).map((el) => el.node as HTMLSpanElement));
+        } else {
+          children.push(node);
+          spans.push(...node.createSpans(start, end));
+        }
+      } else {
+        children.push(node);
+      }
+    }
+    if (wrappableNodes.length) {
+      const spanElement = this.wrapElementsWithSpan(wrappableNodes);
+
+      children.push(spanElement);
+      spans.push(spanElement.node);
+    }
+
+    this.children = children;
+    return spans;
+  }
+
+  removeSpans(spans: HTMLSpanElement[]) {
+    for (let i = this.children.length - 1; i >= 0; i--) {
+      const el = this.children[i];
+
+      if (el instanceof DDSpanElement) {
+        if (spans.includes(el.node)) {
+          el.removeNode();
+
+          this.children.splice(i,1, ...el.children);
+        } else {
+          el.removeSpans(spans);
+        }
+      }
+    }
+
+    let stack: DDTextElement[] = [];
+    const result = [];
+    const checkStack = () => {
+      if (stack.length > 0) {
+        const mainElement = stack[0];
+
+        if (stack.length > 1) {
+          const extraElements = stack.slice(1);
+
+          mainElement.mergeWith(extraElements);
+          extraElements.forEach(el => el.removeNode());
+        }
+
+        result.push(mainElement);
+        stack = [];
+      }
+    };
+
+    for (const el of this.children) {
+      if (el instanceof DDTextElement && (stack.length === 0 || stack[stack.length - 1].node.nextSibling === el.node)) {
+        stack.push(el);
+      } else {
+        checkStack();
+        result.push(el);
+      }
+    }
+    checkStack();
+    this.children = result;
+  }
+}
+
+class DDSpanElement extends DDBlock {
+  public node: HTMLSpanElement;
+  constructor(node: HTMLSpanElement, start: number, end: number) {
+    super(start, end);
+    this.node = node;
+  }
+  removeNode() {
+    const { node } = this;
+    const doc = node.ownerDocument;
+    const parent = node.parentNode as Node;
+    const fragment = doc.createDocumentFragment();
+
+    while (node.firstChild) {
+      fragment.appendChild(node.firstChild);
+    }
```

Comment:
Can you add a comment here about what this is doing? I am not sure how this works at a glance and can only infer that appendChild mutates the value in some way that firstChild is iterated until null.

Reply:
That's exactly as you said. `appendChild` changes parent of `firstChild` wich moves it from it's previous parent's children list and changing of this list changes `firstChild` also untill all childrend leave the parent.